File: go/core/registry.go
package core

import "fmt"

// Step registry — port of registry.ts / registry.rs. Persistent-value
// semantics: AddStep / DefineParameterType return a new Registry; the argument
// is unchanged.

// FormatFn is a parameter-type display formatter (the inverse of parse): it
// renders a value back in the document's notation. ok=false → fall through to
// the generic rendering chain.
type FormatFn func(v Value) (string, bool)

// CustomParameterType is a custom parameter type as registered by an author —
// name plus bare pattern source (the string the registry artifact serializes).
type CustomParameterType struct {
	Name   string
	Regexp string
}

// StepRegistration is one registered step: source expression, source location,
// handler, compiled expression, and role (Kind may be nil — the kindless path).
type StepRegistration struct {
	Expression           string
	ExpressionSourceFile string
	ExpressionSourceLine int
	Handler              Handler
	Kind                 *StepKind
	compiled             *compiledExpression
}

// Registry is the step registry.
type Registry struct {
	Steps                []*StepRegistration
	CustomParameterTypes []CustomParameterType
	customParse          map[string]ParseFn
	Formats              map[string]FormatFn
}

// CreateRegistry returns an empty registry.
func CreateRegistry() Registry {
	return Registry{
		Steps:                nil,
		CustomParameterTypes: nil,
		customParse:          map[string]ParseFn{},
		Formats:              map[string]FormatFn{},
	}
}

func (r Registry) cloneCustomParse() map[string]ParseFn {
	m := make(map[string]ParseFn, len(r.customParse))
	for k, v := range r.customParse {
		m[k] = v
	}
	return m
}

func (r Registry) cloneFormats() map[string]FormatFn {
	m := make(map[string]FormatFn, len(r.Formats))
	for k, v := range r.Formats {
		m[k] = v
	}
	return m
}

// AddStep compiles expression against registry's parameter types and appends it,
// returning a new Registry. Errors on a duplicate expression or an
// un-compilable one.
func AddStep(registry Registry, expression, sourceFile string, sourceLine int, handler Handler, kind *StepKind) (Registry, error) {
	for _, existing := range registry.Steps {
		if existing.Expression == expression {
			return registry, &RegistryError{
				Kind: ErrDuplicateStep,
				Message: fmt.Sprintf(
					"duplicate step definition for %q at %s:%d and %s:%d",
					expression,
					existing.ExpressionSourceFile, existing.ExpressionSourceLine,
					sourceFile, sourceLine,
				),
			}
		}
	}
	compiled, err := compileExpression(expression, registry.CustomParameterTypes, registry.customParse)
	if err != nil {
		return registry, err
	}
	steps := make([]*StepRegistration, len(registry.Steps), len(registry.Steps)+1)
	copy(steps, registry.Steps)
	steps = append(steps, &StepRegistration{
		Expression:           expression,
		ExpressionSourceFile: sourceFile,
		ExpressionSourceLine: sourceLine,
		Handler:              handler,
		Kind:                 kind,
		compiled:             compiled,
	})
	return Registry{
		Steps:                steps,
		CustomParameterTypes: registry.CustomParameterTypes,
		customParse:          registry.customParse,
		Formats:              registry.Formats,
	}, nil
}

// DefineParameterType registers a custom parameter type and returns a new
// Registry recording it.
func DefineParameterType(registry Registry, name, regexp string, parse ParseFn) Registry {
	customTypes := make([]CustomParameterType, len(registry.CustomParameterTypes), len(registry.CustomParameterTypes)+1)
	copy(customTypes, registry.CustomParameterTypes)
	customTypes = append(customTypes, CustomParameterType{Name: name, Regexp: regexp})

	parseMap := registry.cloneCustomParse()
	parseMap[name] = parse

	return Registry{
		Steps:                registry.Steps,
		CustomParameterTypes: customTypes,
		customParse:          parseMap,
		Formats:              registry.Formats,
	}
}

// DefineParameterTypeWithFormat is DefineParameterType additionally retaining a
// display format.
func DefineParameterTypeWithFormat(registry Registry, name, regexp string, parse ParseFn, format FormatFn) Registry {
	next := DefineParameterType(registry, name, regexp, parse)
	formats := next.cloneFormats()
	formats[name] = format
	next.Formats = formats
	return next
}

File: rust/core/src/registry.rs
//! Step registry — port of `registry.ts` / `Registry.java`. Wraps the owned
//! [`crate::expression`] layer. Persistent-value semantics: `add_step` /
//! `define_parameter_type` return a new [`Registry`]; the argument is unchanged.

use crate::error::RegistryError;
use crate::expression::{CompiledExpression, ParameterTypeRegistry};
use crate::handler::Handler;
use crate::step_kind::StepKind;
use crate::value::Value;
use std::collections::HashMap;
use std::rc::Rc;

pub use crate::expression::ParseFn;

/// A parameter-type display formatter (the inverse of `parse`): renders a value
/// back in the document's notation. `None` result → fall through to the generic
/// rendering chain.
pub type FormatFn = Rc<dyn Fn(&Value) -> Option<String>>;

/// A custom parameter type as registered by an author — name plus bare pattern
/// source (the string the registry artifact serializes).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CustomParameterType {
    pub name: String,
    pub regexp: String,
}

impl CustomParameterType {
    pub fn new(name: impl Into<String>, regexp: impl Into<String>) -> CustomParameterType {
        CustomParameterType {
            name: name.into(),
            regexp: regexp.into(),
        }
    }
}

/// One registered step: source expression, source location, handler, compiled
/// expression, and role (`kind` may be `None` — the legacy/kindless path).
#[derive(Clone)]
pub struct StepRegistration {
    pub expression: String,
    pub expression_source_file: String,
    pub expression_source_line: usize,
    pub handler: Handler,
    pub compiled: CompiledExpression,
    pub kind: Option<StepKind>,
}

/// The step registry.
#[derive(Clone)]
pub struct Registry {
    pub steps: Vec<Rc<StepRegistration>>,
    pub parameter_types: ParameterTypeRegistry,
    pub custom_parameter_types: Vec<CustomParameterType>,
    pub formats: HashMap<String, FormatFn>,
}

/// An empty registry with a fresh default parameter-type registry. Seeds the
/// display format for the built-in `{emph}` type (its parameter type itself
/// lives in [`ParameterTypeRegistry::new`]); a mismatch renders the value back
/// in single-asterisk emphasis. Byte-identical to the TS port's `seedBuiltins`.
pub fn create_registry() -> Registry {
    let mut formats: HashMap<String, FormatFn> = HashMap::new();
    formats.insert(
        "emph".to_string(),
        Rc::new(|v: &Value| match v {
            Value::String(s) => Some(format!("*{s}*")),
            _ => None,
        }),
    );
    Registry {
        steps: Vec::new(),
        parameter_types: ParameterTypeRegistry::new(),
        custom_parameter_types: Vec::new(),
        formats,
    }
}

/// Compiles `expression` against `registry`'s parameter types and appends it,
/// returning a new [`Registry`]. Errors on a duplicate expression or an
/// un-compilable one.
pub fn add_step(
    registry: &Registry,
    expression: &str,
    expression_source_file: &str,
    expression_source_line: usize,
    handler: Handler,
    kind: Option<StepKind>,
) -> Result<Registry, RegistryError> {
    for existing in &registry.steps {
        if existing.expression == expression {
            return Err(RegistryError::DuplicateStep(format!(
                "duplicate step definition for \"{}\" at {}:{} and {}:{}",
                expression,
                existing.expression_source_file,
                existing.expression_source_line,
                expression_source_file,
                expression_source_line
            )));
        }
    }
    let compiled = CompiledExpression::compile(expression, &registry.parameter_types)
        .map_err(|e| RegistryError::Expression(e.message))?;
    let mut steps = registry.steps.clone();
    steps.push(Rc::new(StepRegistration {
        expression: expression.to_string(),
        expression_source_file: expression_source_file.to_string(),
        expression_source_line,
        handler,
        compiled,
        kind,
    }));
    Ok(Registry {
        steps,
        parameter_types: registry.parameter_types.clone(),
        custom_parameter_types: registry.custom_parameter_types.clone(),
        formats: registry.formats.clone(),
    })
}

/// Registers a custom parameter type and returns a new [`Registry`] recording it.
pub fn define_parameter_type(
    registry: &Registry,
    name: &str,
    regexp: &str,
    parse: ParseFn,
) -> Registry {
    let mut parameter_types = registry.parameter_types.clone();
    parameter_types.define(name, regexp, parse);
    let mut custom_parameter_types = registry.custom_parameter_types.clone();
    custom_parameter_types.push(CustomParameterType::new(name, regexp));
    Registry {
        steps: registry.steps.clone(),
        parameter_types,
        custom_parameter_types,
        formats: registry.formats.clone(),
    }
}

/// As [`define_parameter_type`], additionally retaining a display `format`.
pub fn define_parameter_type_with_format(
    registry: &Registry,
    name: &str,
    regexp: &str,
    parse: ParseFn,
    format: FormatFn,
) -> Registry {
    let mut next = define_parameter_type(registry, name, regexp, parse);
    next.formats.insert(name.to_string(), format);
    next
}
